### ISSUE TITLE
Importing lightning causes large slowdown in dataloading with pickle files

### ISSUE THREAD

**Issue by lebrice** [Lightning-AI/lightning#18152]
That issue is caused by `legacy_import_unpickler.py` that is setting `pickle.Unpickler = RedirectingUnpickler`.
For datasets that unpickle files, switching from the C unpickler to the Python unpickler creates a slowdown of `~4x`.

Here's minimal code to reproduce (minus of course the paths):
```python
import inspect
import time

# import lightning
import torch
import tqdm
from torch.utils.data import Dataset


class SlowDataset(Dataset):
    def __init__(self):
        super().__init__()

    def __len__(self):
        return 1000

    def __getitem__(self, idx):
        fname = f"/tmp/data/ckpt.pt"
        file = torch.load(fname)
        return file


def main():
    train_dataset = SlowDataset()

    for i, data in enumerate(tqdm.tqdm(train_dataset)):
        if i > 1000:
            break


if __name__ == "__main__":
    main()
```

cc @tchaton @justusschock @awaelchli @borda

**Comment by patrickmineault** [comment]
The slowdown can be triggered by importing `lightning`, and it can be removed by overriding the Unpickler as `legacy_import_unpickler.py`, so it's definitely the RedirectingUnpickler that causes the issue:

```
old_unpickler = pickle.Unpickler
import lightning

pickle.Unpickler = old_unpickler
```

**Comment by awaelchli** [comment]
I saved random data into a `ckpt.pt` file to run the provided code, but I'm not able to show a difference in iteration speed, even when removing the patched pickler. Could you let me know the precise version of lightning, torch, and what you have saved into ckpt.pt?

**Comment by patrickmineault** [comment]
Lightning 2.0.5, torch 2.0.1. The checkpoint contains pickled classes, which might cause the problem. I've included pstats files generated by the profiler which shows the issue - you can see that the python-based pickle gets called all the way down to unpickle classes in the case of lightning. 
[pstats.zip](https://github.com/Lightning-AI/lightning/files/12152848/pstats.zip)

**Comment by awaelchli** [comment]
@patrickmineault Does this PR here fix the issue for you? https://github.com/Lightning-AI/lightning/pull/18154 
To test it, you can run
```
pip install git+https://github.com/Lightning-AI/lightning@bugfix/redirect-unpickler
```

It's removing the global patch of the unpickler. 
If you could help confirm it, that would be very helpful.

**Comment by patrickmineault** [comment]
Yes, that does fix it, thank you!

**Comment by awaelchli** [comment]
I just merged the fix #18154. The added test should ensure it stays this way.
Please consider installing Lightning from source until the commit is added to the next release. 
Thank you for providing a reproducible code snippet and helping me confirm the fix.
